### DISCUSSION TITLE
Multiple connection issues on Linux

### DISCUSSION THREAD

**Discussion by thunder-weasel** [stlehmann/pyads#422] [Q&A] [answered]
Multiple simultaneous instances of the following code works flawlessly on Windows:
```
plc_ip = '10.0.0.4'
plc_ams_id = '10.0.0.4.1.1'

our_port = pyads.open_port()

with pyads.Connection(plc_ams_id, pyads.PORT_TC3PLC1, plc_ip) as plc:
    plc.open()
    while True:
        print(plc.read_by_name('IO.tick1ms'))
        sleep(1)
```
A single instance works fine on Linux, however, when a second invocation is started the first instanced fails with:
```
53606
54609
55613
2022-10-20T10:26:30-0500 Info: connection closed by remote
2022-10-20T10:26:31-0500 Error: write frame failed with error: 32
Traceback (most recent call last):
  File "/mnt/store/vm/vshare/development/python3/beckhoff/./bare_test_01.py", line 15, in <module>
    print(plc.read_by_name('IO.tick1ms'))
  File "/home/jpiel/.local/lib/python3.10/site-packages/pyads/connection.py", line 540, in read_by_name
    return adsSyncReadByNameEx(
  File "/home/jpiel/.local/lib/python3.10/site-packages/pyads/pyads_ex.py", line 1154, in adsSyncReadByNameEx
    handle = adsGetHandle(port, address, data_name)
  File "/home/jpiel/.local/lib/python3.10/site-packages/pyads/pyads_ex.py", line 890, in adsGetHandle
    handle = adsSyncReadWriteReqEx2(
  File "/home/jpiel/.local/lib/python3.10/site-packages/pyads/pyads_ex.py", line 774, in adsSyncReadWriteReqEx2
    raise ADSError(err_code)
pyads.pyads_ex.ADSError: ADSError: Unknown Error (-1).
```
Any tips or ideas on what I'm doing wrong would be greatly appreciated.

**Comment by thunder-weasel** [comment]
Additional info: Above script runs fine with multiple instances on FreeBSD 13.1.

**Reply by tboegi** [reply]
For Linux, you can try 
https://github.com/birkenfeld/ads_forwarder_rs

(However, I wonder why it works under FreeBSD for you ?
 Do you have TwinCAT installed on this system ?)

**Comment by chrisbeardy** [comment]
You're not the first person to bring this up, we believe it may be an issue with the underlying ADS library, so there is no easy fix in pyads. The way around it for now, if I remember correctly, is instead of running the program multiple times, is to run one program and create multiple instances of the connection class. This however I believe only works if you plan on accessing it from just the single PC, i believe if you have a connection from a linux pc, then open one from a windows pc for example, you would get this error still.

**Comment by chrisbeardy** [comment]
linked to #256 and #317

**Comment by chrisbeardy** [comment, marked as answer]
@thunder-weasel see #314 if tthat helps

**Comment by thunder-weasel** [comment]
"...i believe if you have a connection from a linux pc, then open one from a windows pc for example, you would get this error still."  This is not the case.  I just tested with 3 instances of the script running on a Windows virtual machine at the same time as 3 instances running on the FreeBSD PLC.  A single instance still worked fine on Linux, but that instance dies when a second one is started.  I did, also, start a 4th instance on both Windows and BSD and that had no effect on the Linux instance.

I tend to concur that this is probably a Linux issue.  When I have a moment I'll repeat this exercise with C, if the results are the same I'll start working with my Beckhoff contacts to see if it can be resolved.  Thank you much for the help and for your work on pyads!

**Comment by thunder-weasel** [comment]
Testing Notes and response I got from Beckhoff:

> Two separate scripts to two different targets -> Works!
> Two separate scripts to the same target -> Only the last executing script works, the others fail at -1 error
> Single script connecting to two different targets -> Fails to connect on the second Open() command with ADSError: ROUTERERR_PORTALREADYINUSE
> I think it's a limitation of the PyADS on linux, primarily due to it using our adslib.so
> If pyAds was using our new .Net core sample, it would work.
> Alternatively, he could use our C# .Net Core API and get that functionality
> 
> Not sure if this is helpful, but this could describe why a Windows Client doesn’t have issues, it internally must be wrapping our Windows ADS dll…
> 
> Hopefully, the pyAds community can reach out and better explain what is happening under the hood.

**Comment by metalika80** [comment]
I don't think using anything .Net on linux is an option.
Anyways, the workaround is tricking ADS into thinking that each connection is to a different plc.
It's explained in issue 49: https://github.com/Beckhoff/ADS/issues/49
It's not the most elegant solution, but it works. I have 5 connections to the same plc without problems.

**Comment by chrisbeardy** [comment]
@thunder-weasel could you please reclose this issue to help us manage the issue tracker. I don't believe this is an issue with pyads itself but rather Beckhoffs Linux ADS so best to close it here. Thanks.

**Comment by fengyc** [comment]
try this https://github.com/fengyc/ads-proxy

**Comment by AytajHsnv** [comment]
Hallo I used this solution but it didnt work for me. i used the code which you put for multiprocessing solution but different addresses, it throw an error. 
<AmsAddress 10.79.128.68.1.2:30000>
<AmsAddress 10.79.128.68.1.2:30000>
<AmsAddress 10.79.128.68.1.2:30001>
<AmsAddress 10.79.128.68.1.2:30001>
2026-07-29T16:02:45+0200 Error: read frame failed with error: Connection reset by peer
2026-07-29T16:02:45+0200 Info: connection closed by remote
67
2026-07-29T16:02:48+0200 Info: connection closed by remote
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/polysecure/projects/Multi_Detect/utilities/test_plc_time.py", line 39, in connect_and_read
    print(plc.read_by_name("GVL_Python.nPlcTime"))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/polysecure/multidetect/lib/python3.12/site-packages/pyads/connection.py", line 547, in read_by_name
    plc_datatype = self._query_plc_datatype_from_name(data_name,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/polysecure/multidetect/lib/python3.12/site-packages/pyads/connection.py", line 185, in _query_plc_datatype_from_name
    info = adsGetSymbolInfo(self._port, self._adr, data_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/polysecure/multidetect/lib/python3.12/site-packages/pyads/pyads_ex.py", line 928, in adsGetSymbolInfo
    symbol_info = adsSyncReadWriteReqEx2(
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/polysecure/multidetect/lib/python3.12/site-packages/pyads/pyads_ex.py", line 790, in adsSyncReadWriteReqEx2
    raise ADSError(err_code)
pyads.pyads_ex.ADSError: ADSError: timeout elapsed (1861).

**Reply by metalika80** [reply]
On Linux, this solution works for sure: https://github.com/Beckhoff/ADS/issues/49. I used it for a while, but in the end, I switched to router  solution that worked in the same way of tricking the ADS into thinking that each connection is to a different plc, because it was easier to scale across multiple projects.
In your case, this are the terminal commands, assuming enp1s0 is your network interface, 10.79.128.69 is unused, 10.79.128.21 is your linux IP and 10.79.128.31 is unused :
sudo iptables -t nat -A OUTPUT -d 10.79.128.69 -j DNAT --to-destination 10.79.128.68
sudo ip address add 10.79.128.21/24 dev enp1s0
sudo ip route add 10.79.128.31/24 dev enp1s0 src 10.79.128.21/24

**Reply by chrisbeardy** [reply]
it looks like for some reason the ADS is not creating unique ports on your setup. 

<AmsAddress 10.79.128.68.1.2:30000>
<AmsAddress 10.79.128.68.1.2:30000>
<AmsAddress 10.79.128.68.1.2:30001>
<AmsAddress 10.79.128.68.1.2:30001>

These should be unique, you could have the same issue I had with mulitpocessing on when I tried on a Raspberry Pi.  Though its is interesting 2 are unique (two pairs). Maybe the multiprocessing is not working, did you try the threading or just declaring multiple instances of the connection class?

I would try the above setup setting ip tables etc or use https://github.com/fengyc/ads-proxy as a test.
